### ISSUE TITLE
SQL error on pages using module after upgrade to 2020-05-12 release

### ISSUE THREAD

**Issue by bgenere** [cosmocode/dokuwiki-plugin-struct#499]
It seems a new column is now required with name **rid** in a structure table.

All my structure tables does not include this field as they were created before this release. 
Is this a bug ?
Is there any script to upgrade the tables ?

See here the current effects on my web site :
https://notreasso.encom1.fr/doku.php/sommaire
https://notreasso.encom1.fr/doku.php/wiki/manifestations/evenements/sommaire
Thanks for your help.

**Comment by splitbrain** [comment]
Your tables should have been upgraded automatically by the sqlite plugins migration mechanism. Do you have the latest sqlite plugin?

Can you issue a "SELECT * FROM opts" on the struct database in the sqlite admin interface and post the result here?

**Comment by bgenere** [comment]
I get this result line
dbversion | 16

**Comment by splitbrain** [comment]
Hmm. That is the correct version. Let's try something. Make a backup of your `data/meta/struct.sqlite3` file. Then issue the following SQL statement:

```
UPDATE opts SET val=15 WHERE opt='dbversion'
```

This should restart the migration on the next page load.

**Comment by bgenere** [comment]
Done and the issue is solved.
The pid field has been created when the main site page was loaded.

For your information I have done the upgrade using the plugin manager and did not get any error during the update. I made update for many plugins none having problems. The issue appears when I get back to the main page.
I tried to used configuration manager to clean up cache but this was not solving the issue either.

Tomorrow I will check on another web site where the module is not yet updated. 

Thanks a lot for your quick response.

**Comment by splitbrain** [comment]
In what order did you upgrade struct and sqlite? I think the problem might happen with an outdated sqlite.

**Comment by bgenere** [comment]
I do all updates in the same plugin manager session and in alphabetical order (as the list appears),
so SQLlite was updated before Struct plugin.

**Comment by bgenere** [comment]
Before the update of plugins I update Dokuwiki to the Release 2018-04-22c "Greebo" (from b)
That's why I update plugins actually to prevent any mismatch.

**Comment by stvoigt** [comment]
I had the same problem... I updated sqlite first, too. The SQL statement mentioned above fixed it.

**Comment by PifPof73** [comment]
Same issues on two different wikis (both under Ubuntu).
The displayed Verison is v17, resetting to v15 helps, but only in the preview.

Is there a caching problem additionally?

**Comment by ebora** [comment]
I had the same issue, dbversion was 17 when applying the suggested fix, which solved the issue completely. Thank you!

**Comment by alexanderbeermann** [comment]
I am having the same issue.
I updated first the struct plugin - this results in an error. Afterwards I updated the struct plugin and accordingly the struct plugin. This returned in error 500. With the DokuWiki Recovery script my wiki is running again. 

As soon as I activate struct plugin, i receive in the frontend the following errors:
`HY000 10 disk I/O error:
INSERT INTO data_process SELECT *, 0 FROM temp_data_process` 

The command "SELECT * FROM opts" results in "dbversion = 15"
Afterwards I issued command "UPDATE opts SET val=15 WHERE opt='dbversion'" this results in "0 affected rows in 0.00182 seconds"

If I reactivate the struct plugin now, I am facing the additional error "`Das 'Struct Plugin' benötigt das 'Sqlite Plugin'. Bitte installieren und aktivieren."`
I doubt that the migration has not been processed properly.

Any tips how to solve this are appreciated.
KInd regards
Alex

**Comment by annda** [comment]
@alexanderbeermann The first error message you posted seems to indicate a more general problem with writing to the database. It is not directly related to the migration, but it prevents it from executing. Can you start troubleshooting by checking permissions for the database file `data/meta/struct.sqlite3` ?

The second error message is sometimes displayed when migrations fail. I am investigating the problem.

**Comment by alexanderbeermann** [comment]
Hi @annda, 
thank you very much for these hints. 
I changed the permission of `data/meta/struct.sqlite3` from `0640` to `0644`.
The latest changes to the sqlite db have been recently before I upgraded to Hogfather version. 
So, you might be true that something is broken with the connection to the db.

Via the SQLite admin console it is possible to issue select statements from the db.
Is it plausible that I damaged something with the recovery script? I am already missing some assignments of permissions...
Where do I have to correct these permissions.

Kind regards

Kind regards
Alex

**Comment by alexanderbeermann** [comment]
Hello, 
in parallel I asked my hosting provider, if he is able to see from the logs, what is going wrong.
With a small test script they were able to create a new SQlite tabel, insert several datasets, delete them and finally dropped the table,
==> Thus means SQLite is running in general

Finally they proposed me to add the following code to /conf/dokuwiki.php
`putenv('TMPDIR=/is/htdocs/user_tmp/wp1234567_RNEY0LDRR3/');`

I can confirm, the plugins SQLite and STRUCT are working again, but with less performance then before. 
Is this variable reasonable?
Why was Dokuwiki with both plugins workingbefore?

Kind regards
Alex

**Comment by annda** [comment]
@alexanderbeermann Thanks for the update. So it is definitely a system problem, not related to struct.

I don't think setting environment variables in the application this way is a long term solution. First of all, it will be overwritten by DokuWiki upgrade. Your provider should make sure that PHP has reliable access to temporary storage.

**Comment by impactfactory** [comment]
Updated wiki to Release 2023-04-04a "Jack Jackrum". Then I updatet every plugin alphabetically down the line. When I updated the Struct Plugin, I got the total error as seen in the image.

Deinstalled, installed with Interface. Did the same again manually. Then tried "UPDATE opts SET val=17 WHERE opt='dbversion'", also with 15, on struct table. No way.

The data seems to be still, but I cannot install struct. Im glad for a way out.

<img width="562" alt="Bildschirm­foto 2023-07-17 um 17 51 19" src="https://github.com/cosmocode/dokuwiki-plugin-struct/assets/72793580/f2d54e31-d37c-4218-a4fd-520b53e1201a">
